module/zfs/fm.c

```diff
@@ -471,10 +471,19 @@ zfs_zevent_init(zfs_zevent_t **zep)
 void
 zfs_zevent_destroy(zfs_zevent_t *ze)
 {
-	mutex_enter(&zevent_lock);
-	if (ze->ze_zevent)
-		list_remove(&ze->ze_zevent->ev_ze_list, ze);
-	mutex_exit(&zevent_lock);
+
+	/*
+	 * In the event of a module shutdown, all of the zevents have already
+	 * been drained and we're just cleaning up.  Avoid taking the
+	 * zevent_lock here because it may already be destroyed by the time we
+	 * are observing it.
+	 */
+	if ((zevent_flags & ZEVENT_SHUTDOWN) == 0) {
+		mutex_enter(&zevent_lock);
+		if (ze->ze_zevent)
+			list_remove(&ze->ze_zevent->ev_ze_list, ze);
+		mutex_exit(&zevent_lock);
+	}
```

Comment:
Perhaps this special handling should instead be placed in the `sysevent_worker()` thread around the `zfs_zevent_destroy()` call (or some other mechanism).  It looks to me like all other callers of the `zevent_lock` must be initialized.  Let's see if we can get @freqlabs or @amotin's thoughts on this.

Reply:
No objection either way... we need the below kmem_free(), wasn't sure if we needed another new zfs_zevent_* KPI for just that (the conflation in this namespace of those operating on zevent_t vs. zfs_zevent_t is a bit confusing to me) or if the caller can just kmem_free directly instead.

Reply:
I am thinking `kmem_free` in `sysevent_worker`, with a comment explaining the race we need to avoid.